pkg/virt-controller/watch/node.go

```diff
@@ -165,65 +163,193 @@ func (c *NodeController) execute(key string) error {
 		}
+	}
 
-		vmis = filterStuckVirtualMachinesWithoutPods(vmis, pods)
-
-		errs := []string{}
-		// Do sequential updates, we don't want to create update storms in situations where something might already be wrong
-		for _, vmi := range vmis {
-			c.recorder.Event(vmi, v1.EventTypeNormal, NodeUnresponsiveReason, fmt.Sprintf("virt-handler on node %s is not responsive, marking VMI as failed", vmi.Status.NodeName))
-			logger.V(2).Infof("Moving vmi %s in namespace %s on unresponsive node to failed state", vmi.Name, vmi.Namespace)
-			phasePatch := fmt.Sprintf(`{ "op": "replace", "path": "/status/phase", "value": "%s" }`, virtv1.Failed)
-			operation := "add"
-			if vmi.Status.Reason != "" {
-				operation = "replace"
-			}
-			reasonPatch := fmt.Sprintf(`{ "op": "%s", "path": "/status/reason", "value": "%s" }`, operation, NodeUnresponsiveReason)
-			_, err := c.clientset.VirtualMachineInstance(vmi.Namespace).Patch(vmi.Name, types.JSONPatchType, []byte(fmt.Sprintf("[%s, %s]", phasePatch, reasonPatch)))
-			if err != nil {
-				errs = append(errs, fmt.Sprintf("failed to move vmi %s in namespace %s to final state: %v", vmi.Name, vmi.Namespace, err))
-				logger.Reason(err).Errorf("Failed to move vmi %s in namespace %s to final state", vmi.Name, vmi.Namespace)
-			}
-		}
+	if len(errs) > 0 {
+		return fmt.Errorf("%v", strings.Join(errs, "; "))
+	}
 
-		if len(errs) > 0 {
-			return fmt.Errorf("%v", strings.Join(errs, "; "))
-		}
+	return nil
+}
+
+func (c *NodeController) createAndApplyPatch(vmi *virtv1.VirtualMachineInstance, logger *log.FilteredLogger) error {
+	c.recorder.Event(vmi, v1.EventTypeNormal, NodeUnresponsiveReason, fmt.Sprintf("virt-handler on node %s is not responsive, marking VMI as failed", vmi.Status.NodeName))
+	logger.V(2).Infof("Moving vmi %s in namespace %s on unresponsive node to failed state", vmi.Name, vmi.Namespace)
+
+	patch := generateVMIPatch(vmi.Status.Reason)
+	_, err := c.clientset.VirtualMachineInstance(vmi.Namespace).Patch(vmi.Name, types.JSONPatchType, patch)
+	if err != nil {
+		logger.Reason(err).Errorf("Failed to move vmi %s in namespace %s to final state", vmi.Name, vmi.Namespace)
+		return err
 	}
-	if nodeExists {
-		c.Queue.AddAfter(key, c.recheckInterval)
+
+	return nil
+}
+
+func generateVMIPatch(reason string) []byte {
+	phasePatch := fmt.Sprintf(`{ "op": "replace", "path": "/status/phase", "value": "%s" }`, virtv1.Failed)
+	operation := "add"
+	if reason != "" {
+		operation = "replace"
+	}
+	reasonPatch := fmt.Sprintf(`{ "op": "%s", "path": "/status/reason", "value": "%s" }`, operation, NodeUnresponsiveReason)
+	return []byte(fmt.Sprintf("[%s, %s]", phasePatch, reasonPatch))
+}
+
+func (c *NodeController) requeueIfExists(key string, node *v1.Node) {
+	if node == nil {
+		return
+	}
+
+	c.Queue.AddAfter(key, c.recheckInterval)
+}
+
+func (c *NodeController) markNodeAsUnresponsive(node *v1.Node, logger *log.FilteredLogger) error {
```

Comment:
perhaps it's a good place to log here as well, so whoever is debugging this will be able to get the info from one log.

Reply:
Good point, thank you for the review! this has been added.